# Midterm, Part 2
Your midterm has two parts. Make sure you answered the part 1. In this part, you are going to use Ozone dataset. This part tests your ability to use the provided hints and guidelines to perform data analysis on Ozone dataset. Once you're done, upload this notebook on Camino under the Assignment called "Midterm, Part 2" the same way you've been uploading your lab solutions. - Nicholas Fong

### Step 1: Load the python libraries you will need for this project

In [1]:
import os
import numpy as np
import pandas as pd
import csv
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
import statsmodels.formula.api as smf
from sklearn import feature_selection, linear_model
from mpl_toolkits.mplot3d import Axes3D

pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

%matplotlib inline
plt.style.use('ggplot')

### Step 2: Read in your data set

In [2]:
df = pd.read_csv(os.path.join('ozone (1).csv'))
df

Ozone  Solar.R  Wind  Temp  Month  Day
0     41.0    190.0   7.4    67      5    1
1     36.0    118.0   8.0    72      5    2
2     12.0    149.0  12.6    74      5    3
3     18.0    313.0  11.5    62      5    4
4      NaN      NaN  14.3    56      5    5
..     ...      ...   ...   ...    ...  ...
148   30.0    193.0   6.9    70      9   26
149    NaN    145.0  13.2    77      9   27
150   14.0    191.0  14.3    75      9   28
151   18.0    131.0   8.0    76      9   29
152   20.0    223.0  11.5    68      9   30

[153 rows x 6 columns]

## Questions
#### Question 1. Explore the dataset and decide which variables suffer from missing data

In [3]:
# hint: what is the data 'length'?
print len(df)
print df['Ozone'].isnull().sum()
print df['Solar.R'].isnull().sum()
print df['Wind'].isnull().sum()
print df['Temp'].isnull().sum()
print df['Month'].isnull().sum()
print df['Day'].isnull().sum()

153
37
7
0
0
0
0


In [4]:
df.describe()

Ozone     Solar.R        Wind        Temp       Month         Day
count  116.000000  146.000000  153.000000  153.000000  153.000000  153.000000
mean    42.129310  185.931507    9.957516   77.882353    6.993464   15.803922
std     32.987885   90.058422    3.523001    9.465270    1.416522    8.864520
min      1.000000    7.000000    1.700000   56.000000    5.000000    1.000000
25%     18.000000  115.750000    7.400000   72.000000    6.000000    8.000000
50%     31.500000  205.000000    9.700000   79.000000    7.000000   16.000000
75%     63.250000  258.750000   11.500000   85.000000    8.000000   23.000000
max    168.000000  334.000000   20.700000   97.000000    9.000000   31.000000

Answer: Ozone and Solar.R suffer missing data

#### Question 2. Drop rows that have missing values in ALL the columns you indentified above

Hint: in dropna() if you set how = 'all', it will only drop columns that are suffering from missing values at all varaibles you introduce in subset. If you want to get rid of the row that contains missing values in any of the variables you specify, then you shall set how = 'any'

df.dropna(how = 'all',subset = ['Var1','Var2','Var3'],inplace = True)

The above code will check if all 3 variables specified in df have missing values, if they all have missing values it will drop that row.

In [5]:
df.dropna(how = 'all',subset = ['Ozone','Solar.R'],inplace = True)

In [6]:
df

Ozone  Solar.R  Wind  Temp  Month  Day
0     41.0    190.0   7.4    67      5    1
1     36.0    118.0   8.0    72      5    2
2     12.0    149.0  12.6    74      5    3
3     18.0    313.0  11.5    62      5    4
5     28.0      NaN  14.9    66      5    6
..     ...      ...   ...   ...    ...  ...
148   30.0    193.0   6.9    70      9   26
149    NaN    145.0  13.2    77      9   27
150   14.0    191.0  14.3    75      9   28
151   18.0    131.0   8.0    76      9   29
152   20.0    223.0  11.5    68      9   30

[151 rows x 6 columns]

In [7]:
isnull = df.isnull()
isnull

Ozone Solar.R   Wind   Temp  Month    Day
0    False   False  False  False  False  False
1    False   False  False  False  False  False
2    False   False  False  False  False  False
3    False   False  False  False  False  False
5    False    True  False  False  False  False
..     ...     ...    ...    ...    ...    ...
148  False   False  False  False  False  False
149   True   False  False  False  False  False
150  False   False  False  False  False  False
151  False   False  False  False  False  False
152  False   False  False  False  False  False

[151 rows x 6 columns]

#### Spoiler! If everything is going according to plan you shall be left by 151 observations. Also, it seemed like the first two varibales had missing values. 

#### Question 3. Now make a copy of your dataframe into a dataframe named OzoneImputeMean. Also, please use mean of the variables to fill in missing values in OzoneImputeMean

In [8]:
OzoneImputeMean = df
ozonemean = df.Ozone.mean()
solarmean = df['Solar.R'].mean()
for x in range(len(OzoneImputeMean['Ozone'])):
    if x not in OzoneImputeMean.index:
        continue
    if isnull['Ozone'][x]:
        OzoneImputeMean['Ozone'][x] = ozonemean
for x in range(len(OzoneImputeMean['Solar.R'])):
    if x not in OzoneImputeMean.index:
        continue
    if isnull['Solar.R'][x]:
        OzoneImputeMean['Solar.R'][x] = solarmean

OzoneImputeMean

C:\Users\Nicholas\Anaconda2\lib\site-packages\ipykernel\__main__.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Users\Nicholas\Anaconda2\lib\site-packages\ipykernel\__main__.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


Ozone     Solar.R  Wind  Temp  Month  Day
0    41.00000  190.000000   7.4    67      5    1
1    36.00000  118.000000   8.0    72      5    2
2    12.00000  149.000000  12.6    74      5    3
3    18.00000  313.000000  11.5    62      5    4
5    28.00000  185.931507  14.9    66      5    6
..        ...         ...   ...   ...    ...  ...
148  30.00000  193.000000   6.9    70      9   26
149  42.12931  145.000000  13.2    77      9   27
150  14.00000  191.000000  14.3    75      9   28
151  18.00000  131.000000   8.0    76      9   29
152  20.00000  223.000000  11.5    68      9   30

[151 rows x 6 columns]

In [9]:
#Check if missing values are filled
OzoneImputeMean.describe()

Ozone     Solar.R        Wind        Temp       Month         Day
count  151.000000  151.000000  151.000000  151.000000  151.000000  151.000000
mean    42.129310  185.931507    9.941722   78.165563    7.019868   15.801325
std     28.884028   88.544727    3.524984    9.198138    1.406984    8.832531
min      1.000000    7.000000    1.700000   57.000000    5.000000    1.000000
25%     21.000000  119.000000    7.400000   73.000000    6.000000    8.000000
50%     42.129310  197.000000    9.700000   79.000000    7.000000   16.000000
75%     46.500000  257.000000   11.500000   85.000000    8.000000   23.000000
max    168.000000  334.000000   20.700000   97.000000    9.000000   31.000000

#### Question 4. You're a data scientist. Follow the following guidelines and write necessary codes to predict a few variables.

In [10]:
#### Before we start let's define dummy variables for variable Month - don't worry about day! 
# hints: 
# first use panda's get_dummies()
month_dummy =  pd.get_dummies(df.Month, prefix = 'Month')
# then delete Month_9 variable
month_dummy = month_dummy.drop('Month_9', 1)
# then concatenate the dummies to the original data OzoneData
OzoneImputeMean = OzoneImputeMean.join([month_dummy])
# finally run the this line:
OzoneImputeMean.head()

Ozone     Solar.R  Wind  Temp  Month  Day  Month_5  Month_6  Month_7  \
0   41.0  190.000000   7.4    67      5    1      1.0      0.0      0.0   
1   36.0  118.000000   8.0    72      5    2      1.0      0.0      0.0   
2   12.0  149.000000  12.6    74      5    3      1.0      0.0      0.0   
3   18.0  313.000000  11.5    62      5    4      1.0      0.0      0.0   
5   28.0  185.931507  14.9    66      5    6      1.0      0.0      0.0   

   Month_8  
0      0.0  
1      0.0  
2      0.0  
3      0.0  
5      0.0

In [11]:
# now let's explore correlation Matrix
OzoneImputeMean.corr()

Ozone   Solar.R      Wind      Temp     Month       Day   Month_5  \
Ozone    1.000000  0.302970 -0.534163  0.630583  0.151089 -0.011472 -0.281111   
Solar.R  0.302970  1.000000 -0.055581  0.271990 -0.073886 -0.147120 -0.023842   
Wind    -0.534163 -0.055581  1.000000 -0.466032 -0.175317  0.042365  0.237781   
Temp     0.630583  0.271990 -0.466032  1.000000  0.397427 -0.136876 -0.637816   
Month    0.151089 -0.073886 -0.175317  0.397427  1.000000 -0.007727 -0.702257   
Day     -0.011472 -0.147120  0.042365 -0.136876 -0.007727  1.000000  0.011003   
Month_5 -0.281111 -0.023842  0.237781 -0.637816 -0.702257  0.011003  1.000000   
Month_6 -0.065820  0.023896  0.046054  0.050753 -0.362131 -0.017044 -0.242766   
Month_7  0.251524  0.175960 -0.144638  0.318103 -0.007201  0.011471 -0.247805   
Month_8  0.264054 -0.073214 -0.166105  0.321680  0.355246  0.011471 -0.247805   

          Month_6   Month_7   Month_8  
Ozone   -0.065820  0.251524  0.264054  
Solar.R  0.023896  0.175960 -0.073214  
Wind     0.046054 -0.144638 -0.166105  
Temp     0.050753  0.318103  0.321680  
Month   -0.362131 -0.007201  0.355246  
Day     -0.017044  0.011471  0.011471  
Month_5 -0.242766 -0.247805 -0.247805  
Month_6  1.000000 -0.253080 -0.253080  
Month_7 -0.253080  1.000000 -0.258333  
Month_8 -0.253080 -0.258333  1.000000

####  What seems to be the list of best variables can define response variable Ozone? how about response variable Solar.R?

Answer:  Ozone ~ Temperature, Solar.R, Wind     Solar.R ~ Ozone, Temp

In [12]:
# now let's use a regression model to predict Ozone. First drop NaN values in Ozone and save it in 
# OzoneDroppedValues_Ozone
# then run a regression line on variables of interest and check significancy of your model.
# In particular use these variables ['Solar.R','Wind','Temp','Month_5','Month_6','Month_7','Month_8'] 
# to predict
OzoneDroppedVales_Ozone = OzoneImputeMean.Ozone.isnull()
from sklearn.linear_model import LinearRegression
from sklearn import feature_selection

def linear_modeling_with_sklearn(X, y):
    model = linear_model.LinearRegression(fit_intercept = True)
    model.fit(X, y)

    print 'F-statistic (performed for each regressor independently)'
    print '- F-value', feature_selection.f_regression(X, y)[0]
    print '- p-value', feature_selection.f_regression(X, y)[1]
    print 'R^2 =', model.score(X, y)
    print 'Coefficients'
    print '- beta_0 (intercept) =', model.intercept_
    print '- beta_n (n > 0)     =', model.coef_
    
X = OzoneImputeMean[['Solar.R','Wind','Temp','Month_5','Month_6','Month_7','Month_8'] ]
y = OzoneDroppedVales_Ozone

linear_modeling_with_sklearn(X, y)

F-statistic (performed for each regressor independently)
- F-value [ nan  nan  nan  nan  nan  nan  nan]
- p-value [ nan  nan  nan  nan  nan  nan  nan]
R^2 = 1.0
Coefficients
- beta_0 (intercept) = 0.0
- beta_n (n > 0)     = [ 0.  0.  0.  0.  0.  0.  0.]


In [13]:
# finally fill in null values of OzoneData['Ozone'] by predicted values
